src/hdmf/build/classgenerator.py

```diff
@@ -0,0 +1,344 @@
+                                                           not_inherited_fields, type_map)
+                        break  # each field_spec should be processed by only one generator
+
+            for class_generator in self.__custom_generators:
+                class_generator.post_process(classdict, bases, docval_args, spec)
+
+            self.__set_init(classdict, bases, docval_args, not_inherited_fields, spec.name)
+        except TypeDoesNotExistError as e:  # pragma: no cover
+            # this error should never happen after hdmf#322
+            name = spec.data_type_def
+            if name is None:
+                name = 'Unknown'
+            raise ValueError("Cannot dynamically generate class for type '%s'. " % name
+                             + str(e)
+                             + " Please define that type before defining '%s'." % name)
+        cls = ExtenderMeta(data_type, tuple(bases), classdict)
+        return cls
+
+    @classmethod
+    def __set_init(cls, classdict, bases, docval_args, not_inherited_fields, name):
+        # get docval arg names from superclass
+        base = bases[0]
+        parent_docval_args = set(arg['name'] for arg in get_docval(base.__init__))
+        new_args = list()
+        for attr_name, field_spec in not_inherited_fields.items():
+            # auto-initialize arguments not found in superclass
+            if attr_name not in parent_docval_args:
+                new_args.append(attr_name)
+
+        @docval(*docval_args)
+        def __init__(self, **kwargs):
+            if name is not None:  # force container name to be the fixed name in the spec
+                kwargs.update(name=name)
+            pargs, pkwargs = fmt_docval_args(base.__init__, kwargs)
+            base.__init__(self, *pargs, **pkwargs)  # special case: need to pass self to __init__
+
+            for f in new_args:
+                arg_val = kwargs.get(f, None)
+                setattr(self, f, arg_val)
+        classdict['__init__'] = __init__
+
+
+class TypeDoesNotExistError(Exception):  # pragma: no cover
+    pass
+
+
+class CustomClassGenerator:
+    """Subclass this class and register an instance to alter how classes are auto-generated."""
+
+    def __new__(cls, *args, **kwargs):  # pragma: no cover
+        raise TypeError('Cannot instantiate class %s' % cls.__name__)
+
+    # mapping from spec types to allowable python types for docval for fields during dynamic class generation
+    # e.g., if a dataset/attribute spec has dtype int32, then get_class should generate a docval for the class'
+    # __init__ method that allows the types (int, np.int32, np.int64) for the corresponding field.
+    # passing an np.int16 would raise a docval error.
+    # passing an int64 to __init__ would result in the field storing the value as an int64 (and subsequently written
+    # as an int64). no upconversion or downconversion happens as a result of this map
+    _spec_dtype_map = {
```

Comment:
Does this exist anywhere else? Or was it moved here from objectmapper.py?

Reply:
If this exists somewhere else, maybe we should move it to central location and import it from there. No need to do that in this PR though.

Reply:
It was moved from `manager.py`. It's used only in docval generation, so I feel like it makes sense to live in this module if not in this class. There are related but separate mappings in:
1. ObjectMapper: https://github.com/hdmf-dev/hdmf/blob/dev/src/hdmf/build/objectmapper.py#L100
2. spec.DtypeHelper: https://github.com/hdmf-dev/hdmf/blob/dev/src/hdmf/spec/spec.py#L27
3. Validator: https://github.com/hdmf-dev/hdmf/blob/dev/src/hdmf/validate/validator.py#L20
that perhaps we should merge into a central utility module for dtype handling.

Reply:
> perhaps we should merge into a central utility module for dtype handling

Yeah, I think at some point we should do that.